### DISCUSSION TITLE
JSONResponse does not implicitly serialize Pydantic models to JSON

### DISCUSSION THREAD

**Discussion by sam2khat** [fastapi/fastapi#10425] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel
from fastapi.responses import JSONResponse

class Example(BaseModel)
   text: str

app = FastAPI()

@app.get("/{text}")
async def get_root(text):
    example = Example(text=text)
    return JSONResponse(content={"text": example}, status_code=200)

"""
Throws an error upon sending a request -> Object of type 'Example' is not JSON Serializable
"""
```

### Description

When executing the sample code, I get the described error, 

`Object of type 'Example' is not JSON Serializable`

I know FastAPI has already provided its own JSON Encoder [`jsonable_enconder`](https://fastapi.tiangolo.com/tutorial/encoder/) and the documentation also states that it's used internally but apparently the JSONResponse object still uses the native `json.dumps` method to serialize the content. 

Now JSONResponse provided in `fastapi.responses` is actually a direct import from `startlette`. It would be nice if FastAPI could override this functionality to utilize it's own `jsonable_encoder` to serialize the content in JSONResponse

Currently I am using `return JSONResponse(content={"text": jsonable_encoder(example)}, status_code=200)` everywhere

### Operating System

Windows

### FastAPI Version

0.103.1

### Pydantic Version

2.4.0

### Python Version

3.11.4

**Comment by jgould22** [comment, marked as answer]
JSONResponse is really meant to be an escape hatch to allow you to bypass fastapi's internal functionality when required so it would not make sense to add that back in.

In this case though the `Response` classes are pretty simple and you could implement your own response class that does exactly what your example does.

https://github.com/encode/starlette/blob/master/starlette/responses.py#L167

https://fastapi.tiangolo.com/advanced/custom-response/#custom-response-class

**Reply by sam2khat** [reply]
I see. Thanks!

**Comment by gregmuellegger** [comment]
In case you only want to serialize the `example` object and return it in the response you could directly return the object, without needing to wrapping it with `JSONResponse`:

```python
from fastapi import FastAPI
from pydantic import BaseModel

class Example(BaseModel):
   text: str

app = FastAPI()

@app.get("/{text}")
async def get_root(text) -> Example:
    example = Example(text=text)
    return example
```

This results in

```
HTTP/1.1 200 OK
date: Thu, 12 Oct 2023 18:31:47 GMT
server: uvicorn
content-length: 17
content-type: application/json
Connection: close

{
  "text": "mytext"
}
```

**Reply by sam2khat** [reply]
I am aware I can do this, send the Pydantic model directly as a response but as per my requirements, I need to a send JSON response with a status code depending on how the request is handled but thanks for your help! 

I will follow @jgould22's suggestion to create a custom Response class
